### DISCUSSION TITLE
Error message in web ui

### DISCUSSION THREAD

**Discussion by BeatGroove** [raspiblitz/raspiblitz#4227] [Q&A]
Hello,

In the transactions section in the web ui, I get an error that says "500 internal server error".

Any idea how I might fix this?

I noticed this after I was messing around in Zeus, and Zeus crashed on me and gave me some errors. But I don't know if that's related at all.

Since then I have upgraded Raspiblitz from 1.9 to 1.10 and also disabled and reenabled the web ui but the error persists. Everything works fine in Ride The Lightning though and I can see the transactions there.

Thank you.
<img width="600" alt="Screenshot 2023-10-22 at 08 21 13" src="https://github.com/raspiblitz/raspiblitz/assets/148859459/a2d8888f-2f84-43c3-8d61-11cb9ac0a1b4">

**Comment by neversleep404** [comment]
I am experiencing the same issue. It first occurred when I switched from the fat installation to the min installation because CLBOSS could not be installed. I have reinstalled several times, but the web ui problem persists.

**Comment by rootzoll** [comment]
@fusion44 is this a known error

**Reply by fusion44** [reply]
Yes and no. This is a message thrown by the server when something unexpected happened.

**Comment by fusion44** [comment]
@neversleep404 @BeatGroove Can you SSH into your Blitz and search through the logs for errors during the time of when the error appeared?

```sh
sudo journalctl -u blitzapi
```

**Reply by neversleep404** [reply]
Sorry for the late response. I looked into the File. I see an endless error and refreshing the pay_index:

Nov 25 13:00:24 raspberrypi python[1153]: KeyError: 'amount_msat'
Nov 25 13:00:24 raspberrypi python[1153]: 
Nov 25 13:00:24 raspberrypi python[1153]: 2023-11-25 13:00:24 | ❌ | cln_jrpc.py:678 | Got an invoice but could not parse the data: 'amount_msat'
Nov 25 13:00:24 raspberrypi python[1153]: 2023-11-25 13:00:24 | ℹ️ | cln_jrpc.py:967 | Refreshing pay_index
Nov 25 13:00:24 raspberrypi python[1153]: 2023-11-25 13:00:24 | ❌ | cln_jrpc_blitz.py:104 | An error has been caught in function 'list_invoices', process 'MainProcess' (1153), thread '>
Nov 25 13:00:24 raspberrypi python[1153]: Traceback (most recent call last):
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/app/lightning/impl/cln_jrpc.py", line 673, in listen_invoices
Nov 25 13:00:24 raspberrypi python[1153]:     i = Invoice.from_cln_json(data)
Nov 25 13:00:24 raspberrypi python[1153]:         │       │             └ {'label': 'keysend-1699837071.985357627', 'bolt11': 'lnbc1pj4zay0sp5404tpuuxxqrfsx848g632tw9s26tvs5e69je409uva7>
Nov 25 13:00:24 raspberrypi python[1153]:         │       └ <classmethod(<function Invoice.from_cln_json at 0x7fff1fdaafc0>)>
Nov 25 13:00:24 raspberrypi python[1153]:         └ <class 'app.lightning.models.Invoice'>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/app/lightning/models.py", line 706, in from_cln_json
Nov 25 13:00:24 raspberrypi python[1153]:     amt = parse_cln_msat(i["amount_msat"])
Nov 25 13:00:24 raspberrypi python[1153]:           │              └ {'label': 'keysend-1699837071.985357627', 'bolt11': 'lnbc1pj4zay0sp5404tpuuxxqrfsx848g632tw9s26tvs5e69je409uva7pfw82>
Nov 25 13:00:24 raspberrypi python[1153]:           └ <function parse_cln_msat at 0x7fff1fda8400>
Nov 25 13:00:24 raspberrypi python[1153]: KeyError: 'amount_msat'
Nov 25 13:00:24 raspberrypi python[1153]: During handling of the above exception, another exception occurred:
Nov 25 13:00:24 raspberrypi python[1153]: Traceback (most recent call last):
Nov 25 13:00:24 raspberrypi python[1153]:   File "<frozen runpy>", line 198, in _run_module_as_main
Nov 25 13:00:24 raspberrypi python[1153]:   File "<frozen runpy>", line 88, in _run_code
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/venv/lib/python3.11/site-packages/uvicorn/__main__.py", line 4, in <module>
Nov 25 13:00:24 raspberrypi python[1153]:     uvicorn.main()
Nov 25 13:00:24 raspberrypi python[1153]:     │       └ <Command main>
Nov 25 13:00:24 raspberrypi python[1153]:     └ <module 'uvicorn' from '/home/blitzapi/blitz_api/venv/lib/python3.11/site-packages/uvicorn/__init__.py'>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/venv/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
Nov 25 13:00:24 raspberrypi python[1153]:     return self.main(*args, **kwargs)
Nov 25 13:00:24 raspberrypi python[1153]:            │    │     │       └ {}
Nov 25 13:00:24 raspberrypi python[1153]:            │    │     └ ()
Nov 25 13:00:24 raspberrypi python[1153]:            │    └ <function BaseCommand.main at 0x7fff21930c20>
Nov 25 13:00:24 raspberrypi python[1153]:            └ <Command main>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/venv/lib/python3.11/site-packages/click/core.py", line 1055, in main
Nov 25 13:00:24 raspberrypi python[1153]:     rv = self.invoke(ctx)
Nov 25 13:00:24 raspberrypi python[1153]:          │    │      └ <click.core.Context object at 0x7fff226ab150>
Nov 25 13:00:24 raspberrypi python[1153]:          │    └ <function Command.invoke at 0x7fff21930040>
Nov 25 13:00:24 raspberrypi python[1153]:          └ <Command main>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/venv/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
Nov 25 13:00:24 raspberrypi python[1153]:     return ctx.invoke(self.callback, **ctx.params)
Nov 25 13:00:24 raspberrypi python[1153]:            │   │      │    │           │   └ {'port': 11111, 'host': '0.0.0.0', 'root_path': '/api', 'app': 'app.main:app', 'uds': None, 'fd': >
Nov 25 13:00:24 raspberrypi python[1153]:            │   │      │    │           └ <click.core.Context object at 0x7fff226ab150>
Nov 25 13:00:24 raspberrypi python[1153]:            │   │      │    └ <function main at 0x7fff218554e0>
Nov 25 13:00:24 raspberrypi python[1153]:            │   │      └ <Command main>
Nov 25 13:00:24 raspberrypi python[1153]:            │   └ <function Context.invoke at 0x7fff219316c0>
Nov 25 13:00:24 raspberrypi python[1153]:            └ <click.core.Context object at 0x7fff226ab150>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/venv/lib/python3.11/site-packages/click/core.py", line 760, in invoke
Nov 25 13:00:24 raspberrypi python[1153]:     return __callback(*args, **kwargs)
Nov 25 13:00:24 raspberrypi python[1153]:                        │       └ {'port': 11111, 'host': '0.0.0.0', 'root_path': '/api', 'app': 'app.main:app', 'uds': None, 'fd': None, 'reloa>
Nov 25 13:00:24 raspberrypi python[1153]:                        └ ()
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/venv/lib/python3.11/site-packages/uvicorn/main.py", line 404, in main
Nov 25 13:00:24 raspberrypi python[1153]:     run(
Nov 25 13:00:24 raspberrypi python[1153]:     └ <function run at 0x7fff219c7100>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/venv/lib/python3.11/site-packages/uvicorn/main.py", line 569, in run
Nov 25 13:00:24 raspberrypi python[1153]:     server.run()
Nov 25 13:00:24 raspberrypi python[1153]:     │      └ <function Server.run at 0x7fff219c6ac0>
Nov 25 13:00:24 raspberrypi python[1153]:     └ <uvicorn.server.Server object at 0x7fff21834810>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/venv/lib/python3.11/site-packages/uvicorn/server.py", line 60, in run
Nov 25 13:00:24 raspberrypi python[1153]:     return asyncio.run(self.serve(sockets=sockets))
Nov 25 13:00:24 raspberrypi python[1153]:            │       │   │    │             └ None
Nov 25 13:00:24 raspberrypi python[1153]:            │       │   │    └ <function Server.serve at 0x7fff219c6b60>
Nov 25 13:00:24 raspberrypi python[1153]:            │       │   └ <uvicorn.server.Server object at 0x7fff21834810>
Nov 25 13:00:24 raspberrypi python[1153]:            │       └ <function run at 0x7fff21dfcb80>
Nov 25 13:00:24 raspberrypi python[1153]:            └ <module 'asyncio' from '/usr/lib/python3.11/asyncio/__init__.py'>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/usr/lib/python3.11/asyncio/runners.py", line 190, in run
Nov 25 13:00:24 raspberrypi python[1153]:     return runner.run(main)
Nov 25 13:00:24 raspberrypi python[1153]:            │      │   └ <coroutine object Server.serve at 0x7fff2186c480>
Nov 25 13:00:24 raspberrypi python[1153]:            │      └ <function Runner.run at 0x7fff21bd8400>
Nov 25 13:00:24 raspberrypi python[1153]:            └ <asyncio.runners.Runner object at 0x7fff21867510>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/usr/lib/python3.11/asyncio/runners.py", line 118, in run
Nov 25 13:00:24 raspberrypi python[1153]:     return self._loop.run_until_complete(task)
Nov 25 13:00:24 raspberrypi python[1153]:            │    │     │                  └ <Task pending name='Task-1' coro=<Server.serve() running at /home/blitzapi/blitz_api/venv/lib/python>
Nov 25 13:00:24 raspberrypi python[1153]:            │    │     └ <function BaseEventLoop.run_until_complete at 0x7fff21bd2020>
Nov 25 13:00:24 raspberrypi python[1153]:            │    └ <_UnixSelectorEventLoop running=True closed=False debug=False>
Nov 25 13:00:24 raspberrypi python[1153]:            └ <asyncio.runners.Runner object at 0x7fff21867510>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/usr/lib/python3.11/asyncio/base_events.py", line 640, in run_until_complete
Nov 25 13:00:24 raspberrypi python[1153]:     self.run_forever()
Nov 25 13:00:24 raspberrypi python[1153]:     │    └ <function BaseEventLoop.run_forever at 0x7fff21bd1f80>
Nov 25 13:00:24 raspberrypi python[1153]:     └ <_UnixSelectorEventLoop running=True closed=False debug=False>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/usr/lib/python3.11/asyncio/base_events.py", line 607, in run_forever
Nov 25 13:00:24 raspberrypi python[1153]:     └ <_UnixSelectorEventLoop running=True closed=False debug=False>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/usr/lib/python3.11/asyncio/base_events.py", line 607, in run_forever
Nov 25 13:00:24 raspberrypi python[1153]:     self._run_once()
Nov 25 13:00:24 raspberrypi python[1153]:     │    └ <function BaseEventLoop._run_once at 0x7fff21bd3d80>
Nov 25 13:00:24 raspberrypi python[1153]:     └ <_UnixSelectorEventLoop running=True closed=False debug=False>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/usr/lib/python3.11/asyncio/base_events.py", line 1922, in _run_once
Nov 25 13:00:24 raspberrypi python[1153]:     handle._run()
Nov 25 13:00:24 raspberrypi python[1153]:     │      └ <function Handle._run at 0x7fff21daa3e0>
Nov 25 13:00:24 raspberrypi python[1153]:     └ <Handle Task.task_wakeup(<Future finis... ...}, ...]}}>)>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/usr/lib/python3.11/asyncio/events.py", line 80, in _run
Nov 25 13:00:24 raspberrypi python[1153]:     self._context.run(self._callback, *self._args)
Nov 25 13:00:24 raspberrypi python[1153]:     │    │            │    │           │    └ <member '_args' of 'Handle' objects>
Nov 25 13:00:24 raspberrypi python[1153]:     │    │            │    │           └ <Handle Task.task_wakeup(<Future finis... ...}, ...]}}>)>
Nov 25 13:00:24 raspberrypi python[1153]:     │    │            │    └ <member '_callback' of 'Handle' objects>
Nov 25 13:00:24 raspberrypi python[1153]:     │    │            └ <Handle Task.task_wakeup(<Future finis... ...}, ...]}}>)>
Nov 25 13:00:24 raspberrypi python[1153]:     │    └ <member '_context' of 'Handle' objects>
Nov 25 13:00:24 raspberrypi python[1153]:     └ <Handle Task.task_wakeup(<Future finis... ...}, ...]}}>)>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/app/lightning/service.py", line 236, in _handle_invoice_listener
Nov 25 13:00:24 raspberrypi python[1153]:     async for i in ln.listen_invoices():
Nov 25 13:00:24 raspberrypi python[1153]:                    │  └ <function LnNodeCLNjRPCBlitz.listen_invoices at 0x7fff1f32a2a0>
Nov 25 13:00:24 raspberrypi python[1153]:                    └ <app.lightning.impl.specializations.cln_jrpc_blitz.LnNodeCLNjRPCBlitz object at 0x7fff1ff0b150>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/app/lightning/impl/specializations/cln_jrpc_blitz.py", line 176, in listen_invoices
Nov 25 13:00:24 raspberrypi python[1153]:     async for i in super().listen_invoices():
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/app/lightning/impl/cln_jrpc.py", line 681, in listen_invoices
Nov 25 13:00:24 raspberrypi python[1153]:     await self._refresh_invoice_sub(True)
Nov 25 13:00:24 raspberrypi python[1153]:           │    └ <function LnNodeCLNjRPC._refresh_invoice_sub at 0x7fff1f329080>
Nov 25 13:00:24 raspberrypi python[1153]:           └ <app.lightning.impl.specializations.cln_jrpc_blitz.LnNodeCLNjRPCBlitz object at 0x7fff1ff0b150>
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/app/lightning/impl/cln_jrpc.py", line 969, in _refresh_invoice_sub
Nov 25 13:00:24 raspberrypi python[1153]:     invoices = await self.list_invoices(
Nov 25 13:00:24 raspberrypi python[1153]:                      │    └ <function LnNodeCLNjRPCBlitz.list_invoices at 0x7fff1f3298a0>
Nov 25 13:00:24 raspberrypi python[1153]:                      └ <app.lightning.impl.specializations.cln_jrpc_blitz.LnNodeCLNjRPCBlitz object at 0x7fff1ff0b150>
Nov 25 13:00:24 raspberrypi python[1153]: > File "/home/blitzapi/blitz_api/app/lightning/impl/specializations/cln_jrpc_blitz.py", line 104, in list_invoices
Nov 25 13:00:24 raspberrypi python[1153]:     return await super().list_invoices(
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/app/lightning/impl/cln_jrpc.py", line 307, in list_invoices
Nov 25 13:00:24 raspberrypi python[1153]:     invoices.append(Invoice.from_cln_json(i))
Nov 25 13:00:24 raspberrypi python[1153]:     │        │      │       │             └ {'label': 'keysend-1699837071.985357627', 'bolt11': 'lnbc1pj4zay0sp5404tpuuxxqrfsx848g632tw9s26tvs5>
Nov 25 13:00:24 raspberrypi python[1153]:     │        │      │       └ <classmethod(<function Invoice.from_cln_json at 0x7fff1fdaafc0>)>
Nov 25 13:00:24 raspberrypi python[1153]:     │        │      └ <class 'app.lightning.models.Invoice'>
Nov 25 13:00:24 raspberrypi python[1153]:     │        └ <method 'append' of 'list' objects>
Nov 25 13:00:24 raspberrypi python[1153]:     └ []
Nov 25 13:00:24 raspberrypi python[1153]:   File "/home/blitzapi/blitz_api/app/lightning/models.py", line 706, in from_cln_json
Nov 25 13:00:24 raspberrypi python[1153]:     amt = parse_cln_msat(i["amount_msat"])
Nov 25 13:00:24 raspberrypi python[1153]:           │              └ {'label': 'keysend-1699837071.985357627', 'bolt11': 'lnbc1pj4zay0sp5404tpuuxxqrfsx848g632tw9s26tvs5e69je409uva7pfw82>
Nov 25 13:00:24 raspberrypi python[1153]:           └ <function parse_cln_msat at 0x7fff1fda8400>
Nov 25 13:00:24 raspberrypi python[1153]: KeyError: 'amount_msat'

**Reply by neversleep404** [reply]
this occurs a many times per second. That could also be a reason for my high CPU load

**Reply by neversleep404** [reply]
ChatGPT suggest:
The error you're encountering is a `KeyError` related to the `'amount_msat'` key in your Python application. This kind of error typically occurs when you try to access a key in a dictionary that doesn't exist. In your case, it seems that the code is trying to access the `'amount_msat'` key in an invoice object, but this key is not present in the dictionary.

Here's a step-by-step approach to resolve this issue:

1. **Check for Optional Fields**: The `'amount_msat'` field might not always be present in every invoice. It's essential to check whether this field exists before trying to access it. You can use the `.get()` method of dictionaries in Python, which allows you to provide a default value if the key is not found. For example:
   ```python
   amt = parse_cln_msat(i.get("amount_msat", default_value))
   ```
   Replace `default_value` with an appropriate default value for your context.

2. **Data Validation**: Ensure that the data you're receiving (in this case, invoices) contains all the necessary fields. If the `'amount_msat'` field is essential, your data validation step should catch invoices that do not have this field.

3. **Error Handling**: Modify the error handling in your code to manage missing keys more gracefully. This could involve logging a warning and skipping the current invoice processing or providing a fallback mechanism.

4. **Review Data Source**: If the `'amount_msat'` field is supposed to be always present, there might be an issue with how the data is being generated or fetched. Review the source of your invoice data to ensure it's consistent with your application's expectations.

5. **Update Documentation/Code Comments**: If the `'amount_msat'` field is optional, update your code documentation or comments to reflect this, so it's clear to anyone else working with the code.

6. **Test with Different Scenarios**: After making changes, test your application with various invoice data, including cases where the `'amount_msat'` field is missing, to ensure your changes handle all scenarios correctly.

By implementing these steps, you should be able to resolve the `KeyError` and handle similar issues more effectively in the future.

**Reply by onlinecrash** [reply]
Hello, 
I have exactly the same error.
Is there a solution for the 'problem'?
I can't do much with the ChatGPT output.
I would be very grateful for a tip!
Many thanks in advance!

**Reply by neversleep404** [reply]
I ignored the problem and have not used the web UI since then.

**Reply by onlinecrash** [reply]
Is there anything I can do to solve the problem?
Or will this be fixed with an update at some point?
I found this article:
https://github.com/raspiblitz/raspiblitz/issues/4298

**Reply by neversleep404** [reply]
As you see, it is not fixed.
(i am now offline for a few days)
